runtime/interpreter/interpreter_transaction.go

```diff
@@ -148,6 +173,74 @@ func (interpreter *Interpreter) declareTransactionEntryPoint(declaration *ast.Tr
 }
 
 func (interpreter *Interpreter) VisitTransactionRoleDeclaration(_ *ast.TransactionRoleDeclaration) StatementResult {
-	// TODO:
-	panic("TODO")
+	panic(errors.NewUnreachableError())
+}
+
+func (interpreter *Interpreter) declareTransactionRole(
+	declaration *ast.TransactionRoleDeclaration,
+) (
+	roleValue *SimpleCompositeValue,
+	prepareFunctionValue *HostFunctionValue,
+) {
+	transactionRoleType := interpreter.Program.Elaboration.TransactionRoleDeclarationType(declaration)
+
+	var prepareFunction *ast.FunctionDeclaration
+	var prepareFunctionType *sema.FunctionType
+	if declaration.Prepare != nil {
+		prepareFunction = declaration.Prepare.FunctionDeclaration
+		prepareFunctionType = transactionRoleType.PrepareFunctionType()
+	}
+
+	staticType := NewCompositeStaticTypeComputeTypeID(
+		interpreter,
+		interpreter.Location,
+		fmt.Sprintf(".%s", declaration.Identifier.Identifier),
+	)
+
+	roleValue = NewSimpleCompositeValue(
+		interpreter,
+		staticType.TypeID,
+		staticType,
+		nil,
+		map[string]Value{},
+		nil,
+		nil,
+		nil,
+	)
+
+	// Construct a raw HostFunctionValue without a type,
+	// instead of using NewHostFunctionValue, which requires a type.
+	//
+	// This host function value is an internally created and used function,
+	// and can never be passed around as a value.
+	// Hence, the type is not required.
+
+	common.UseMemory(interpreter, common.HostFunctionValueMemoryUsage)
+
+	prepareFunctionValue = &HostFunctionValue{
+		Function: func(invocation Invocation) Value {
+			interpreter.activations.PushNewWithCurrent()
+			defer interpreter.activations.Pop()
+
+			self := MemberAccessibleValue(roleValue)
+			invocation.Self = &self
+			interpreter.declareVariable(sema.SelfIdentifier, self)
+
+			transactionRoleScope := interpreter.activations.CurrentOrNew()
+
+			if prepareFunction != nil {
```

Comment:
Would it be more efficient to early-exit at the top of this function? It might save additional activation push/pop, and self-var declarations.

Reply:
I left it in the same format as the interpretation of transactions because it is the same as for transactions. Currently, role blocks only have a prepare, but they'll likely get support for conditions like in transactions. f3f15b714e45bfc9cca9814e5e6e3054230de236 improves naming a bit